returnn/tf/layers/base.py

```diff
@@ -1271,17 +1271,23 @@ def batch_norm(self, data,
       else:
         sample_variance = updated_sample_variance
       # If train or if force_sample, use default use_sample=0.0, otherwise use_sample=1.0.
-      if self.network.train_flag is not False or force_sample:
-        if force_sample:
-          pass  # leave use_sample as-is
-        else:
-          use_sample = tf.where(self.network.train_flag, use_sample, 1.0)
+      skip_updates = self.network.train_flag is False and update_sample_only_in_training and not force_sample
+      if skip_updates:
+        mean = sample_mean
+        variance = sample_variance
       else:
-        use_sample = 1.0
-      mean = (1. - use_sample) * mean + use_sample * sample_mean
-      variance = (1. - use_sample) * variance + use_sample * sample_variance
-      bn = (data.placeholder - mean) * tf_compat.v1.rsqrt(variance + epsilon)
-      if delayed_ops:
+        if self.network.train_flag is not False or force_sample:
+          if force_sample:
+            pass  # leave use_sample as-is
+          else:
+            use_sample = tf.where(self.network.train_flag, use_sample, 1.0)
+        else:
+          use_sample = 1.0
+        mean = (1. - use_sample) * mean + use_sample * sample_mean
+        variance = (1. - use_sample) * variance + use_sample * sample_variance
+      variance_with_epsilon = variance if epsilon == 0.0 else variance + epsilon
```

Comment:
We have `opt_add` for that. And I personally would prefer to keep the code shorter and do it directly in the line `bn = ...`.